### ISSUE TITLE
CsvHealthReporter output is unbounded

### ISSUE THREAD

**Issue by karolz-ms** [Azure/diagnostics-eventflow#138]
We should have a limit on the output of CsvHealthReporter (e.g. employ log rotation).

Also, while working in this codebase, double-check that transient write failures in the reporter do not bring the whole EventFlow pipeline down.

**Comment by xiaomi7732** [comment]
Based on the discussion with @karolz-ms, there are 3 issues in total:
**Issue one** is easy, it is the root cause of #139 - that it crashed the pipeline when write permission is not here for the health report; I'll confirm this issue and put in a fix to fail gracefully.

**Issue two** is an unlimited sized health report: Although we have been split the csv health report by date, there still could be a huge amount of log data generated in one day.
My proposal for this is to introduce a maximum file size in the configuration file so that a single health report file won't exceed the size of the file.
```csharp
"healthReporter": {
    "type": "CsvHealthReporter",
    "logFilePrefix": "HealthReport",
// ...
    "singleLogFileMaximumSizeInMBytes": "8192"  // Default to 8G
}

```
`The default file size will be 8GB`. When the size of the file exceeds the maximum, we will reset the file stream, start the writing from 0 again. To balance the frequency of the check and the performance, I also suggest we check the size of the file every 30 seconds rather than every event. This will might lead to file bleeds the top size of the settings but shall bring performance gains when a huge number of events happens.

**Issue three** is over time, the total log file count/log size will become too big for the system to continuously working.
Since there is no auto cleaning/deleting mechanism, the log file number/size, in theory, is unlimited, which will eventually blow the storage.
How about we introduce a retention policy:
```csharp
"healthReporter": {
    "type": "CsvHealthReporter",
    "logFilePrefix": "HealthReport",
...
    "singleLogFileMaximumSizeInMBytes": "8192"
    "RententionLogsInDays": "30"  // Default value to 30 days.
}
```
We will retain the log for 30 days by default. The user could control the days of retention through configuration. The deletion could happen using the same trigger of switching days of the health report file name by days.

@karolz-ms ,
What do you think? Are we on the same page? :-)

**Comment by karolz-ms** [comment]
I like the proposal a lot, thank you @xiaomi7732!

I would suggest one simple (?) refinement though: instead of having just one file per day, let's have one or two. The reporter would start writing first file. If the size limit is exceeded, the other file is started. If that file also exceeds the limit, the first file is truncated to 0 and becomes the active file. From this time the files are reused in a alternating fashion until new day comes.

This way you lose at most half of the old data when the log file hit the limit. It is like log rotation with just two log files.

**Comment by xiaomi7732** [comment]
@karolz-ms,
I considered 2 file rotation as well. But isn't it keeps more data if we have only 1 file, we re-write of from the beginning but keep all existing data there?
For example, we want to write 0 to 9 into the log, 5 digits are the limit.  After got 12345:
12345 -> ~~1~~**6**2345 -> 6~~2~~**7**345 -> 67~~3~~**8**45 -> 678~~4~~**9**5
v.s.:
(123)(45) -> (~~123~~6)(45) -> (67)(45) -> (678)(45) -> (678)(~~45~~9)

**Comment by xiaomi7732** [comment]
@karolz-ms & I have a discussion offline, we come down to a conclusion that it'd be clearer to have a two-log file rotation than one. I'll put that into implementation.
At the same time, I understand the two of us can't predicate every case. If anyone has any concern, please feel free to raise it. :-)
If I hear no other comment, I'll start to put the implementation in and send the PR when I have it.

**Comment by karolz-ms** [comment]
Go for it! :-)

**Comment by xiaomi7732** [comment]
On the record, I did some benchmark on how to get the file size. There I benched 3 ways to get the size of the file:
* Use local variable
* Use FileStream.Position
* FileInfo.Size

I benched 5 iterations for each method and for each iteration, it will write a small message for 100000000 times. Intentionally, I try to read the position of the file stream per write.
At this point, I am leaning toward using FileStream.Position to track the file size:

* Local variable and FileStream.Position has almost the same result;
* FileInfo.Size has a very similar result when the file is small but after the file becomes big (A few hundreds of MB), the time consumption increases dramatically.
* Since the FileStream.Position will be ensured by the .NET Fx to have an accurate evaluation for writing the logs with very little overhead v.s. local variable, I'll try to make use of that property than tracking the size using another variable. That will also simplify the case when there are multiple threads that a mismatch might happen.

Benchmark Details:
-
| Method,(100000000 times writing) | 1         | 2         | 3         | 4         | 5         | Average    |
|----------------------------------|-----------|-----------|-----------|-----------|-----------|------------|
| Local,variable                   | 5.6312926 | 6.4342356 | 6.5329318 | 6.6786149 | 6.607974  | 6.37700978 |
| FileStream.Position              | 6.4457961 | 6.4672116 | 6.6030152 | 6.4295214 | 6.6829105 | 6.52569096 |
| FileInfo.Size                    | x         | x         | x         | x         | x         | x          |

**FileInfo.Size** took too much time to finish 100000000 writes.

Code used:
```csharp
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

namespace BenchMarkStreamPosition
{
    class Program
    {
        public static int Dictionray { get; private set; }

        static void Main(string[] args)
        {
            Program p = new Program();

            Dictionary<string, Func<byte[], int, long>> dictionary = new Dictionary<string, Func<byte[], int, long>>() {
                { "Local variable", p.WriteStream },
                { "FileStream.Position", p.WriteStream1},
                // { "FileInfo.Size", p.WriteStream2},
            };

            for (int i = 0; i < 5; i++)
            {
                Console.Write($"Round: {i} - ");
                foreach (var benchmarkItem in dictionary)
                {
                    // Set benchmark target;
                    Func<byte[], int, long> func = benchmarkItem.Value;
                    Console.WriteLine($"{benchmarkItem.Key}.");

                    // Start bench mark
                    byte[] content = Encoding.UTF8.GetBytes("This is just a test string");
                    long position = 0;

                    // Start benchmark
                    const int iterations = 100000000;

                    Stopwatch watch = Stopwatch.StartNew();
                    position = func(content, iterations);
                    watch.Stop();
                    Console.Write($"End Position: {position}.");
                    Console.WriteLine($"Time Elapsed: {watch.Elapsed}.");
                    Console.WriteLine($"============== Finished {benchmarkItem.Key} round {i} ================");
                }
            }
        }

        long WriteStream(byte[] content, int iterations)
        {
            using (FileStream fileStream = new FileStream("c:\\temp\\1.log", FileMode.Append))
            {
                long pos = fileStream.Position;
                for (int i = 0; i < iterations; i++)
                {
                    fileStream.Write(content, 0, content.Length);
                    pos += content.Length;
                }
                return pos;
            }
        }

        long WriteStream1(byte[] content, int iteration)
        {
            using (FileStream fileStream = new FileStream("c:\\temp\\2.log", FileMode.Append))
            {
                long pos = fileStream.Position;
                for (int i = 0; i < iteration; i++)
                {
                    fileStream.Write(content, 0, content.Length);
                    // For this funciton, we cloud direclty return pos after the loop finishes. However, we do want to see the perf
                    // to check the position every write.
                    pos = fileStream.Position;
                }
                return pos;
            }
        }

        long WriteStream2(byte[] content, int iteration)
        {
            using (FileStream fileStream = new FileStream("c:\\temp\\3.log", FileMode.Append))
            {
                long pos = fileStream.Position;
                for (int i = 0; i < iteration; i++)
                {
                    fileStream.Write(content, 0, content.Length);
                    // How bad could this be?
                    pos = new FileInfo("c:\\temp\\3.log").Length;
                }
                return pos;
            }
        }
    }
}
```

**Comment by karolz-ms** [comment]
Makes sense, let's go with FileStream.Position. https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/FileStream.cs#L643

**Comment by karolz-ms** [comment]
Fixed with Core nuget release 1.1.7 or newer
